File: golang/queue.go
package main

type queue struct {
	length      int
	first, last *node
}

func (q *queue) enqueue(item int) {
	node := &node{data: item}
	q.length++
	if q.last == nil {
		q.last = node
		q.last.next = node
		q.first = q.last
		return
	}

	q.last.next = node
	q.last = node

}

func (q *queue) dequeue() int {
	if q.first == nil {
		return -1
	}

	result := q.first.data
	q.first = q.first.next
	q.length--
	if q.length == 0 {
		q.last = nil
	}

	return result
}

func (q *queue) peek() int {
	return q.first.data
}

File: js/src/day1/Queue.ts
type node<T> = {
    value: T;
    next?: node<T>;
};

export default class Queue<T> {
    public length: number;

    private head?: node<T>;
    private tail?: node<T>;

    constructor() {
        this.head = this.tail = undefined;
        this.length = 0;
    }

    enqueue(item: T): void {
        const node: node<T> = { value: item };
        this.length++;
        if (!this.tail) {
            this.tail = node;
            this.tail.next = node;
            this.head = this.tail;
            return;
        }
        this.tail.next = node;
        this.tail = node;
    }

    deque(): T | undefined {
        if (!this.head) {
            return;
        }
        const result = this.head?.value;
        this.head = this.head.next;
        this.length--;
        if (this.length === 0) {
            this.tail = undefined;
        }
        return result;
    }

    peek(): T | undefined {
        return this.tail?.value;
    }
}
